File: leetcode_0783/cpp/leetcode_0783.cpp
//
// Created by wzy on 2022/4/2.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        int min_diff = INT_MAX;
        int pre = -1;
        //方法1.中序遍历(BST中序是单调递增 所以 最小差值一定在相邻节点间)
        inOrderDiff(root,pre,min_diff);
        //方法2.
        //因为是二叉搜索树，所以最小差值必定是左子树的最大值与当前节点 或者 当前节点与右子树最小节点之间
        //这里采用后序遍历寻找最小差值
        // getMinHelper(root,min_diff);
        return min_diff;
    }
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<vector<int>> {};
        if(root)
            cur.push_back(root);
        vector<vector<int>> res;
        vector<int> curEles;
        while(cur.size()){
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                curEles.push_back(node->val);
                if(node->left)
                    next.push_back(node->left);
                if(node->right)
                    next.push_back(node->right);
            }
            res.push_back(curEles);
            curEles.clear();
            cur = next;
            next.clear();
        }
        return res;
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }

private:
    void inOrderDiff(TreeNode* root,int& pre,int& min_diff){
        if(root){
            inOrderDiff(root->left,pre,min_diff);
            //说明是最小节点
            if(pre == -1)
                pre = root->val;
            else{
                min_diff = min(min_diff,abs(root->val - pre));
                pre = root->val;
            }
            inOrderDiff(root->right,pre,min_diff);
        }
    }
    //为了避免反复向下寻找最大节点和最小节点，采用一次性返回子树最大值和最小值
    pair<TreeNode*,TreeNode*> getMinHelper(TreeNode* root,int& min_diff){
        if(root){
            pair<TreeNode*,TreeNode*> leftPair = getMinHelper(root->left,min_diff);
            pair<TreeNode*,TreeNode*> rightPair = getMinHelper(root->right,min_diff);
            if(leftPair.second)
                min_diff = min(abs(root->val - leftPair.second->val),min_diff);
            if(rightPair.first)
                min_diff = min(abs(root->val - rightPair.first->val),min_diff);
            //返回最大最小值节点
            TreeNode* minNode = (leftPair.first)?leftPair.first:root;
            TreeNode* maxNode = (rightPair.second)?rightPair.second:root;
            return make_pair(minNode,maxNode);
        }
        return make_pair(nullptr,nullptr);
    }
};


template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}


int main() {
    TreeNode* root = new TreeNode(236);
    TreeNode* left = new TreeNode(104);
    TreeNode* right = new TreeNode(701);
    TreeNode* ll = nullptr;
    TreeNode* lr = new TreeNode(227);
    TreeNode* rl = nullptr;
    TreeNode* rr = new TreeNode(911);
    root->left = left;
    root->right = right;
    left->left = ll;
    left->right = lr;
    right->left = rl;
    right->right = rr;
    vector<string> vec = Solution().levelOrderStr(root);
    //[236,104,701,null,227,null,911]
    printVec(vec);
    int res = Solution().getMinimumDifference(root);
    cout<<res<<endl;
    //result
    //9
}

File: leetcode_0783/java/leetcode_0783.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_0783 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public int getMinimumDifference(TreeNode root) {
            this.min_diff = Integer.MAX_VALUE;
            //方法1.中序遍历(BST中序是单调递增 所以 最小差值一定在相邻节点间)
            pre = -1;
            inOrderDiff(root);
            //因为是二叉搜索树，所以最小差值必定是左子树的最大值与当前节点 或者 当前节点与右子树最小节点之间
            //这里采用后序遍历寻找最小差值
            // getMinHelper(root);
            return min_diff;
        }
        private int pre;
        private int min_diff;
        private void inOrderDiff(TreeNode root){
            if(root != null){
                inOrderDiff(root.left);
                if(pre == -1)
                    pre = root.val;
                else{
                    min_diff = Math.min(min_diff,Math.abs(root.val-pre));
                    pre = root.val;
                }
                inOrderDiff(root.right);
            }
        }
        private TreeNode[] getMinHelper(TreeNode root){
            if(root!=null){
                TreeNode[] leftPair = getMinHelper(root.left);
                TreeNode[] rightPair = getMinHelper(root.right);
                if(leftPair[1]!=null)
                    min_diff = Math.min(Math.abs(root.val-leftPair[1].val),min_diff);
                if(rightPair[0]!=null)
                    min_diff = Math.min(Math.abs(root.val-rightPair[0].val),min_diff);
                //返回当前最大最小节点
                TreeNode maxNode = (rightPair[1]!=null)?rightPair[1]:root;
                TreeNode minNode = (leftPair[0]!=null)?leftPair[0]:root;
                return new TreeNode[]{minNode,maxNode};
            }
            return new TreeNode[]{null,null};
        }

        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray((T[]) new Object[0])) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_0783.Solution solution = new leetcode_0783().new Solution();
        leetcode_0783.TreeNode root = new leetcode_0783().new TreeNode(236);
        leetcode_0783.TreeNode left = new leetcode_0783().new TreeNode(104);
        leetcode_0783.TreeNode right = new leetcode_0783().new TreeNode(701);
        leetcode_0783.TreeNode ll = null;
        leetcode_0783.TreeNode lr = new leetcode_0783().new TreeNode(227);
        leetcode_0783.TreeNode rl = null;
        leetcode_0783.TreeNode rr = new leetcode_0783().new TreeNode(911);
        left.left = ll;
        left.right = lr;
        right.left = rl;
        right.right = rr;
        root.left = left;
        root.right = right;
//      [236,104,701,null,227,null,911]
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        int res = solution.getMinimumDifference(root);
        System.out.println(res);
        //result
        //9
    }
}
